### ISSUE TITLE
[FEATURE REQUEST] add notification for new stuff

### ISSUE THREAD

**Issue by philippd1** [gymh/GymHerzo-Android#5]
send a notification @ around 07:40 in the morning, when there's something for the current filter

**Comment by Mic-R** [comment]
Because I can't look into what you are using exactly, here is how I would do it.

1. If something in the table changes -> let the sever make a list for every class. After it is finished, set it online.
2. Tell the App to check for it's file at 7:41 (to give the server some time) (according to the filter)
3. If the file is empty => don't send a notification; if file doesn't exist => don't send  notification either; If the App fails => retry for 10 seconds => if it still fails, do not retry, abort (to deal with possible failures)
4. delete the notification at around 11:30 to not cause confusion

**Comment by nicolaiort** [comment]
Sorry Mic that's not how background notifications should be implemented

**Comment by nicolaiort** [comment]
And self-deleting notifications are not really good UX

**Comment by Mic-R** [comment]
> And self-deleting notifications are not really good UX

Yeah, but trust me, if you wouldn't delete it at some point, People will be really confused.

**Comment by Mic-R** [comment]
> Sorry Mic that's not how background notifications should be implemented

Yes, I agree with that. They really shouldn't, but it would work and it would be easy...

**Comment by philippdormann** [comment]
> 1. set it online

not sure what "it" refers to?

> 2. Tell the App to check for it's file (according to the filter) at 7:41 (to give the server some time)

not sure if the check should be performed on client-side or via server-push. I'm pro client-side.
due to Android AlarmManager timings, it will almost never be exactly 07:41:00, so spikes should not be a huge issue.
differing timeouts (e.g. 07:30-07:40) seems like a reasonable suggestion, though.

not sure what is meant by "it's file" 🧐


> 4. delete the notification at around 11:30 to not cause confusion

Yeah, you can just create a background service e.g. AlarmManager that'll have a predefined timeout and proceeds to delete the notification. 
**Whether you "should" actually do that is up for debate. A notification should be there to notify the user... and the user should be able to dismiss it on their own. 🤷‍♂️**

**Comment by Mic-R** [comment]
> > 1. set it online
> 
> not sure what "it" refers to?
That refers to the file, the Server created for every class (5-13 a-f  and so on).




> > 1. Tell the App to check for it's file (according to the filter) at 7:41 (to give the server some time)
> 
> not sure if the check should be performed on client-side or via server-push. I'm pro client-side.
> due to Android AlarmManager timings, it will almost never be exactly 07:41:00, so spikes should not be a huge issue.
> differing timeouts (e.g. 07:30-07:40) seems like a reasonable suggestion, though.

Actually, I get my notification that I scheduled in my self developed calendar (serverside push, you see, I don't like taking up CPU+RAM usage on the clients device) always on time. This could differ on devices, just my personal experiences.
> not sure what is meant by "it's file" 🧐

It's file should mean, the file, that matches with the filter you set.
> > 1. delete the notification at around 11:30 to not cause confusion
> 
> Yeah, you can just create a background service e.g. AlarmManager that'll have a predefined timeout and proceeds to delete the notification.
> **Whether you "should" actually do that is up for debate. A notification should be there to notify the user... and the user should be able to dismiss it on their own. 🤷‍♂️**

I don't know, it was just a thought for the "unexperienced" users, that might get confused.

**Comment by philippdormann** [comment]
**please stop quote replying to entire paragraphs 🥴🥴 that's unreadable.** thanks.

> That refers to the file, the Server created for every class (5-13 a-f and so on).

- There is no file and there will never be any file imo 🤷‍♂️


> I get my notification (serverside push, I don't like taking up CPU+RAM usage on the clients device) always on time.
> This could differ on devices, just my personal experiences.

+ CPU + RAM usage in a daily background job is negligible
- serverside push only works on devices with GPlay Core Services
- serverside push requires server side data persistence (ressources)
- serverside push requires server side scheduling (ressources)

**Comment by Mic-R** [comment]
Ok, not that many/long quote replies, got it.

>There is no file and there will never be any file imo

Ok, thats's ok, it was just the way I would do it, cuz I then I can ensure you, it works.

>CPU + RAM usage in a daily background job is negligible

Yeah, I don't like it that much, but that is just my point of view.

>Serverside push only works on devices with GPlay Core Services

As I know, not many people are still rocking with Android 4.0 or something around that, so that souldn't realy be a problem... (I would be concerned for the people that use Android 4.0) And those people then just are unlucky, they can't use it.

For the rest, ok, I get that, still just the way I would solve it.

**Comment by nicolaiort** [comment]
> Ok, thats's ok, it was just the way I would do it, cuz I then I can ensure you, it works.

You can of course use a file for caching - but using memory cache or serving the requests on the fly is faster and less prone to (random) errors.

> Serverside push only works on devices with GPlay Core Services
> not many people are still rocking with Android 4.0

This does not only concern users with older android versions but also users with devices that have been banned under the US-China tech export war. This affects a lot of devices by Chinese manufacturers such as Huawei.
It may also affect users with rooted devices or custom ROMs.

>  And those people then just are unlucky, they can't use it.

That's not the spirit!
Development should always focus to bring enhancements to as many users as possible

**Comment by Mic-R** [comment]
>This does not only concern users with older android versions but also users with devices that have been banned under the US-China tech export war. This affects a lot of devices by Chinese manufacturers such as Huawei.
It may also affect users with rooted devices or custom ROMs.

I can fully understand your point there. BUt how are those users supposed to install .apk files on their phones? They would need android for that...
Root devices are a security risk, I would fully exvlude them, if I were you.

>That's not the spirit!

Again, Andoid 4.0 is a security risk, what might block other features (like gymag streaming)

**Comment by nicolaiort** [comment]
> They would need android for that

Well they use Android - just without all of the Google stuff.
Either by choice (custom ROM an similar) or by force (trade ban).

**Comment by Mic-R** [comment]
> Well they use Android - just without all of the Google stuff.
> Either by choice (custom ROM an similar) or by force (trade ban).

I can remember something like that but Google is going to switch from .apk to .add oder .abb soon, I am not sure right now.
That would need Google Services and the those devices would not be able to run the app either way, so... does that really mater?

**Comment by Mic-R** [comment]
> > Well they use Android - just without all of the Google stuff.
> > Either by choice (custom ROM an similar) or by force (trade ban).
> 
> I can remember something like that but Google is going to switch from .apk to .add oder .abb soon, I am not sure right now.
> That would need Google Services and the those devices would not be able to run the app either way, so... does that really mater?

It is .AAB

**Comment by philippdormann** [comment]
> That would need Google Services and the those devices would not be able to run the app either way

Android App Bundles using Dynamic Delivery (.aab) are only distributed via Google Play.
As Nicolai and myself already pointed out, Google Play should + **will** not be a requirement.

Installing .xapk, , .apk as well as .apkm files **does not require Google Play**.
**You should do some research before stating anything else.**

Further, device specific APKs can be generated from .aab's

> so... does that really mater?

YES. Yes it does matter.

**Comment by Mic-R** [comment]
>Installing .xapk, , .apk as well as .apkm files does not require Google Play.
>You should do some research before stating anything else.

First of all, to my defense: I did! But I found different information, some saying you won't be able to install .apk any longer and some pointing out, that such app might get flagged by antivirus(es, idk what the plural is) I just forgot to put that in.

>YES. Yes it does matter.

OK. I got it. No need for capslock. So we need to include everyone, I now can get that. So another solution will be needed

**Comment by philippdormann** [comment]
no need to apologize.

please refer to the official Google Article ["New Android App Bundle and target API level requirements in 2021"](https://android-developers.googleblog.com/2020/11/new-android-app-bundle-and-target-api.html) if you're interested in AppBundle requirements

**Comment by Mic-R** [comment]
> please refer to the official Google Article ["New Android App Bundle and target API level requirements in 2021"](https://android-developers.googleblog.com/2020/11/new-android-app-bundle-and-target-api.html) if you're interested in AppBundle requirements

Ok, thanks! That might get important for me in the future. I couldn't find that article on duck duck go (currently not on my main machine).
